### ISSUE TITLE
Error when implementing a new operator: "Check failed: (info.ref_count) == (0) ..."

### ISSUE THREAD

**Issue by nicklhy** [dmlc/mxnet#1729]
Hi~
I am currently implementing a loss layer(or operator in mxnet) which looks quite similar to the triplet loss. The code is written in the C++ level.
I have already finished writing the forward, backward and other necessary functions but got some problems when running it.
"src/symbol/graph_executor.cc:888: Check failed: (info.ref_count) == (0) Because the last operator is not Loss function, head_gradient is required in calling backward."
Technically, it should be a standard loss layer. But by following the softmax-output operator, I didn't output a float value as the overall loss. I choose to output some other things like the probabilities in softmax output. I'm really confused about the error print now. What the ref_count and head_gradient exactly are?
Can any one provide some helps?

**Comment by piiswrong** [comment]
look into softmax_output's DeclareBackwardDependency function. Copy it over and possibly modify it.

**Comment by nicklhy** [comment]
@piiswrong , thanks for your help.
As I request a temp space(xxxOpResource) in forward function and stored some intermediate results there, I do not need any in_data or out_data in the backward phase. And since my operator(layer) should be the last one of the entire graph, I think the out_grad is also not necessary.  Is that right?

**Comment by nicklhy** [comment]
@piiswrong , another question about the forward and backward phase.
I use the following code to test my operator, but it seems that the backward function has been skipped somehow. I put a LOG(INFO) sentence in backward function and nothing being printed out.
`
data = mx.sym.Variable('data')
label = mx.sym.Variable('label')
my_custom = mx.sym.MyCustomOutput(data=data, label=label, some_other_args)
x = mx.nd.empty((10, 100))
label = mx.nd.empty((10, ))
dx = mx.nd.zeros((10, 100))

executor = my_custom.bind(ctx=mx.cpu(), args={'data': x, 'label': y}, args_grad={'x': dx})
executor.forward()
print executor.outputs[0].asnumpy()
executor.backward()
`
The executor.outputs[0] can be printed out with a right value which proves the forward function runs good. But I don't know how the backward goes wrong.

**Comment by piiswrong** [comment]
try adding executor.grad_arrays[0].wait_to_read()

**Comment by piiswrong** [comment]
also, I think temp space is not saved until backward. If you want to pass infromation from forward to backward you should use aux variable or a internal array in the operator.

**Comment by nicklhy** [comment]
@piiswrong , following your advice and I put the intermediate results into a internal Tensor variable in the operator like
```
// Tensor<xpu, 2, real_t> xx; defined as a class variable
xx = NewTensor<xpu, real_t, 2>(shape_xx, 0.0f); // Forward phase
```
But the program just crashed
"An fatal error occurred in asynchronous engine operation. If you do not know what caused this error ... ..."
I'm now wondering if it is prohibited to define a Tensor variable in the operator?

**Comment by nicklhy** [comment]
Thanks a lot. I end up with storing the intermediate results into some variables of the operator class.
